src/ORM/DataList.php

```diff
@@ -817,7 +818,9 @@ public function createDataObject($row)
             $class = $row['RecordClassName'];
         }
 
-        $item = Injector::inst()->create($class, $row, false, $this->getQueryParams());
+        $creationType = empty($row['ID']) ? DataObject::CREATE_OBJECT : DataObject::CREATE_HYDRATED;
+
+        $item = Injector::inst()->create($class, $row, $creationType, $this->getQueryParams());
```

Comment:
> I'm worried about the tri-state for dataobjects, bool state for all other objects and coupling of injector to dataobject mostly.

AFAIK DataObject is the only class that makes use of this 2nd argument, so what we're talking about here is turning a bi-state argument tri-state argument.

Hyrdation, Creation, and Singletons are different cases. In the status-quo code, Creation and Hydration are coupled with the assumption that the Creation case will have an empty record. Frankly, Hydration and Singleton have more in common so you could argue "maybe hydration should set isSingleton = true". But I think that this will risk more messy edge-cases and also risk semver breakages. So I think a tri-state field is better.

We could arguably use the presence of an ID value to distinguish hydration (ID is present) from creation (ID isn't present), but that will cause bugs in two cases:
 * We attempt to hydrate a record without an ID
 * We attempt to create a record with a predefined ID

The former case seems unlikely unless we change our ORM, but the latter could conceivably occur in edge-cases such as copying content from stage to draft. I can't guarantee that it will introduce bugs, but it seems a risky proposition — I think a tri-state is cleaner.

*Why is this conditional needed?*
This condition was added to make ManyManyThroughList::add() work, which creates a _new_ object by calling `$hasManyList->createDataObject()` on line 217.

So, if we wanted to reduce the use of the conditional here, we could potentially deprecate `createDataObject()` in favour of `hydrateDataObject()` and `createNewDataObject()`. I'd probably still leave this code as-is, but at least the messiness would be contain in a deprecated API.

Reply:
I agree with Sam, in that it's safer to make this an explicit decision of the API caller (tri-state), rather than keep the current signature and infer from the various constructor argument combinatoins (bi-state)